From 661767c60b65dea568eba44b1a2057b1efe7e338 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Mon, 21 Feb 2022 02:05:55 +0000
Subject: [PATCH] xen/arm: Apply MPAM config to Dom0

To make it easy to handle MPAM configuration of domains, this commit
firstly introduces `struct mpam_domain` to `struct arch_domain`, which
combines the PARTID and its MPAM config.

Dom0 MPAM config parsed from the Xen cmdline is applied at the boot
time. This commit applies the MPAM config to Dom0 at the boot time by
introducing an interface `mpam_update_domain()` to call
`mpam_update_resources()` to apply the control of MPAM resources.
This function is supposed to be called on all online CPUs because of
the MPAM MSC CPU affinity.

Also, since MPAM normally has architecture maximum width of each
hardware feature and the user configuration cannot exceed this value,
this commit also introduces a helper function `mpam_get_hwinfo()` to
query the architecture maximum width of MPAM class type. This maximum
width is used to sanitise the MPAM config values applied to each domain
by `mpam_sanitise_config()`.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-3395
---
 xen/arch/arm/domain.c             |  10 +++
 xen/arch/arm/domain_build.c       |   6 ++
 xen/arch/arm/include/asm/domain.h |   5 ++
 xen/arch/arm/include/asm/mpam.h   |  16 ++--
 xen/arch/arm/mpam.c               | 134 +++++++++++++++++++++++++++++-
 xen/include/public/arch-arm.h     |   2 +
 6 files changed, 165 insertions(+), 8 deletions(-)

diff --git a/xen/arch/arm/domain.c b/xen/arch/arm/domain.c
index 28e3aaa5e482..1d5a2e7abd08 100644
--- a/xen/arch/arm/domain.c
+++ b/xen/arch/arm/domain.c
@@ -637,6 +637,10 @@ int arch_sanitise_domain_config(struct xen_domctl_createdomain *config)
         }
     }
 
+#ifdef CONFIG_MPAM
+    mpam_sanitise_config(&config->arch.mpam, true);
+#endif
+
     /* The P2M table must always be shared between the CPU and the IOMMU */
     if ( config->iommu_opts & XEN_DOMCTL_IOMMU_no_sharept )
     {
@@ -785,6 +789,12 @@ int arch_domain_create(struct domain *d,
     d->arch.sve_vl = config->arch.sve_vl;
 #endif
 
+#ifdef CONFIG_MPAM
+    d->arch.mpam.cfg.l3_cpbm = config->arch.mpam.l3_cpbm;
+    if ( mpam_enabled )
+        mpam_update_domain(d);
+#endif
+
     return 0;
 
 fail:
diff --git a/xen/arch/arm/domain_build.c b/xen/arch/arm/domain_build.c
index 49792dd590ee..7937cc1bc930 100644
--- a/xen/arch/arm/domain_build.c
+++ b/xen/arch/arm/domain_build.c
@@ -21,6 +21,7 @@
 #include <acpi/actables.h>
 #include <asm/device.h>
 #include <asm/kernel.h>
+#include <asm/mpam.h>
 #include <asm/setup.h>
 #include <asm/tee/tee.h>
 #include <asm/platform.h>
@@ -3915,6 +3916,11 @@ void __init create_dom0(void)
             panic("SVE vector length error\n");
     }
 
+#ifdef CONFIG_MPAM
+    dom0_cfg.arch.mpam.l3_cpbm = dom0_mpam_config.l3_cpbm ?
+                                 dom0_mpam_config.l3_cpbm : 0;
+#endif
+
     dom0 = domain_create(0, &dom0_cfg, CDF_privileged | CDF_directmap);
     if ( IS_ERR(dom0) )
         panic("Error creating domain 0 (rc = %ld)\n", PTR_ERR(dom0));
diff --git a/xen/arch/arm/include/asm/domain.h b/xen/arch/arm/include/asm/domain.h
index 99e798ffff68..1d0e46adc5ec 100644
--- a/xen/arch/arm/include/asm/domain.h
+++ b/xen/arch/arm/include/asm/domain.h
@@ -7,6 +7,7 @@
 #include <asm/p2m.h>
 #include <asm/vfp.h>
 #include <asm/mmio.h>
+#include <asm/mpam.h>
 #include <asm/gic.h>
 #include <asm/vgic.h>
 #include <asm/vpl011.h>
@@ -119,6 +120,10 @@ struct arch_domain
     void *tee;
 #endif
 
+#ifdef CONFIG_MPAM
+    struct mpam_domain mpam;
+#endif
+
 }  __cacheline_aligned;
 
 struct arch_vcpu
diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index bf5f89953c88..24ff85b22491 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -319,12 +319,7 @@ struct mpam_props
 };
 
 #define mpam_has_feature(_feat, x) ((1 << (_feat)) & (x)->features)
-
-static inline void mpam_set_feature(enum mpam_device_features feat,
-                                    struct mpam_props *props)
-{
-    props->features |= (1 << feat);
-}
+#define mpam_set_feature(_feat, x) ((x)->features |= (1 << (_feat)))
 
 static inline void mpam_clear_feature(enum mpam_device_features feat,
                                       mpam_features_t *supported)
@@ -408,10 +403,19 @@ struct mpam_msc_ris
     struct mpam_component *comp;
 };
 
+struct mpam_domain {
+    uint16_t partid;
+    struct xen_arch_domctl_mpam cfg;
+};
+
 extern struct xen_arch_domctl_mpam dom0_mpam_config;
+extern bool mpam_enabled;
 
 extern void mpam_postsmp_setup(void);
 extern void mpam_presmp_setup(void);
+extern void mpam_sanitise_config(struct xen_arch_domctl_mpam *cfg,
+                                 bool domain_create);
+extern void mpam_update_domain(struct domain *d);
 extern int mpam_discovery_cpu_online(unsigned int cpu);
 
 #endif /* __ARM_MPAM_H__ */
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 2186e30b73e8..606a8ae20b25 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -30,6 +30,7 @@
 #include <asm/smp.h>
 
 struct xen_arch_domctl_mpam __initdata dom0_mpam_config;
+bool __read_mostly mpam_enabled = false;
 /*
  * mpam_list_lock protects the global lists when writing.
  * Once the MPAM is enabled these lists are read-only.
@@ -54,8 +55,6 @@ static unsigned int mpam_fw_num_msc;
 static uint16_t mpam_partid_max;
 static uint8_t mpam_pmg_max;
 
-static bool __read_mostly mpam_enabled = false;
-
 static int __init parse_dom0_mpam_cmdline(const char *s)
 {
     char *token, *cmdline = (char *)s;
@@ -1237,6 +1236,69 @@ static int __init mpam_enable(void)
     return 0;
 }
 
+static void mpam_apply_config(struct mpam_component *comp, uint16_t partid,
+                              struct mpam_config *cfg)
+{
+    struct mpam_msc_ris *ris;
+
+    list_for_each_entry(ris, &comp->ris, comp_list)
+    {
+        if ( !cpumask_test_cpu(smp_processor_id(), &ris->affinity) )
+            continue;
+
+        spin_lock(&ris->msc->lock);
+        if ( memcmp(&comp->cfg[partid], cfg, sizeof(*cfg)) )
+            mpam_reprogram_ris_partid(ris, partid, cfg);
+        spin_unlock(&ris->msc->lock);
+    }
+
+    memcpy(&comp->cfg[partid], cfg, sizeof(*cfg));
+}
+
+static void mpam_update_class(struct mpam_class *class, uint16_t partid,
+                              struct mpam_config *mpam_cfg)
+{
+    struct mpam_component *comp;
+
+    list_for_each_entry(comp, &class->components, class_list)
+    {
+        if ( !cpumask_test_cpu(smp_processor_id(), &comp->affinity) )
+            continue;
+
+        mpam_apply_config(comp, partid, mpam_cfg);
+    }
+}
+
+/* This is supposed to be IPI heavy as it should be called on all cpus.. */
+static void mpam_update_resources(void *data)
+{
+    struct mpam_class *class;
+    struct mpam_domain *domain = (struct mpam_domain *)data;
+
+    if ( domain->partid > ACCESS_ONCE(mpam_partid_max) )
+    {
+        printk(XENLOG_ERR
+               "MPAM: PARTID %d exceeds max PARTID supported (%d) in system\n",
+               domain->partid, ACCESS_ONCE(mpam_partid_max));
+        return;
+    }
+
+    list_for_each_entry(class, &mpam_classes, classes_list)
+    {
+        if ( !cpumask_test_cpu(smp_processor_id(), &class->affinity) )
+            continue;
+
+        if ( class->type == MPAM_CLASS_CACHE && class->level == 3 )
+        {
+            struct mpam_config mpam_cfg = {};
+
+            mpam_set_feature(mpam_feat_cpor_part, &mpam_cfg);
+            mpam_cfg.cpbm = domain->cfg.l3_cpbm;
+            mpam_update_class(class, domain->partid, &mpam_cfg);
+        }
+    }
+}
+
 static int __init mpam_msc_probe(void)
 {
     unsigned int num_msc = 0;
@@ -1268,6 +1330,74 @@ static void __init mpam_load_supported_partid_pmg(void)
     ACCESS_ONCE(mpam_pmg_max) = FIELD_GET(MPAMIDR_PMG_MAX, sys_mpam_idr);
 }
 
+/*
+ * Get MPAM hwinfo based on the input resource type.
+ * Currently assume MPAM hwinfo is the same across different clusters.
+ */
+static int mpam_get_hwinfo(enum mpam_class_types type, unsigned int level,
+                           unsigned int *val)
+{
+    struct mpam_class *class;
+
+    list_for_each_entry(class, &mpam_classes, classes_list)
+    {
+        if ( !cpumask_test_cpu(smp_processor_id(), &class->affinity) )
+            continue;
+
+        if ( class->type != type )
+            continue;
+
+        switch (type)
+        {
+        case MPAM_CLASS_CACHE:
+            if ( class->level != level )
+                break;
+            *val = class->props.cpbm_wd;
+            return 0;
+
+        default:
+            printk(XENLOG_ERR "MPAM: Unsupported resource type\n");
+            return -EINVAL;
+        }
+    }
+
+    return -ENOENT;
+}
+
+/*
+ * Sanitise the MPAM config for a domain.
+ * - If given non-empty config, clamp the value to avoid exceeding the system
+ *   max supported value.
+ * - If given an empty config,
+ *   (1) At the domain creation time, use the max supported value as default.
+ *   (2) Otherwise (from xl command), allow the empty value.
+ */
+void mpam_sanitise_config(struct xen_arch_domctl_mpam *cfg, bool domain_create)
+{
+    unsigned int hw_val, hw_lim;
+
+    if ( mpam_get_hwinfo(MPAM_CLASS_CACHE, 3, &hw_val) )
+    {
+        if ( cfg->l3_cpbm )
+            printk(XENLOG_ERR "MPAM: L3 CPBM not supported\n");
+    }
+    else
+    {
+        hw_lim = (1 << hw_val) - 1;
+
+        if ( cfg->l3_cpbm )
+            cfg->l3_cpbm = (cfg->l3_cpbm > hw_lim) ? hw_lim : cfg->l3_cpbm;
+        else
+            cfg->l3_cpbm = domain_create ? hw_lim : cfg->l3_cpbm;
+    }
+}
+
+void mpam_update_domain(struct domain *d)
+{
+    mpam_update_resources(&d->arch.mpam);
+    smp_call_function(mpam_update_resources, &d->arch.mpam, 1);
+}
+
 void __init mpam_presmp_setup(void)
 {
     int rc = 0;
diff --git a/xen/include/public/arch-arm.h b/xen/include/public/arch-arm.h
index 598107e2f756..6f1a4bb642ef 100644
--- a/xen/include/public/arch-arm.h
+++ b/xen/include/public/arch-arm.h
@@ -354,6 +354,8 @@ struct xen_arch_domainconfig {
      *
      */
     uint32_t clock_frequency;
+    /* IN - Contains domain MPAM configuration */
+    struct xen_arch_domctl_mpam mpam;
 };
 #endif /* __XEN__ || __XEN_TOOLS__ */
 
-- 
2.34.1

